File: AjaxVectorObjects/ClientScripts/VectorObjects/RectangleVObject.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
/// <reference name="MicrosoftAjax.js" />
/// <reference path="~/ClientScripts/Math/PointF.js" />
/// <reference path="~/ClientScripts/Math/RectangleF.js" />
/// <reference path="~/ClientScripts/Math/RotatedRectangleF.js" />
/// <reference path="~/ClientScripts/Math/Common.js" />

Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject = function (left, top, width, height) {
    ///	<summary>This client-side class corresponds to the <see cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject" /> server-side class and gives an opportunity to obtain access to its primary members in JavaScript.</summary>
    /// <remarks><para>This class represents a vector rectangle.</para></remarks>
    /// <seealso cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject" />
    /// <constructor>
    /// 	<exclude />
    /// </constructor>
    var ns = Aurigma.GraphicsMill.AjaxControls.VectorObjects;

    if (left != null && top != null && width != null && height != null) {
        ns.RectangleVObject.initializeBase(this, [ns.Math.Path.rectangle(left, top, width, height)]);
        this._controlPoints = [new ns.Math.PointF(left, top), new ns.Math.PointF(left + width, top + height)];
    }
    else
        ns.RectangleVObject.initializeBase(this);

    this._allowNegativeResize = true;
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject.prototype = {
    get_rectangle: function () {
        /// <summary>Gets a rectangle and its rotation angle corresponded to this image vector object.</summary>
        /// <value type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF">The <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> value which specifies a rectangle and its rotation angle.</value>
        var r = Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject.callBaseMethod(this, 'get_rectangle');
        if (r.Width < 0) r.Width = -r.Width;
        if (r.Height < 0) r.Height = -r.Height;
        return r;
    }
};
Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.RectangleVObject", Aurigma.GraphicsMill.AjaxControls.VectorObjects.ShapeVObject);

File: AjaxVectorObjects/VectorObjects/RectangleVObject.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
using System.Drawing.Drawing2D;
using Path = Aurigma.GraphicsMill.AdvancedDrawing.Path;
using RectangleF = System.Drawing.RectangleF;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects
{
    public class RectangleVObject : ShapeVObject
    {
        public RectangleVObject(RectangleF rectangle)
            : this(rectangle.Left, rectangle.Top, rectangle.Width, rectangle.Height)
        {
        }

        public RectangleVObject()
            : this(0, 0, MinDimension, MinDimension)
        {
        }

        public RectangleVObject(float left, float top, float width, float height)
        {
            Path = Math.Path.CreateRectanglePath(left, top, width, height);
        }

        protected internal bool IncludeBorder = false;

        public override RectangleF Bounds
        {
            get
            {
                if (!IncludeBorder)
                    return base.Bounds;

                var rectangle = base.Bounds;
                rectangle.X -= BorderWidth / 2;
                rectangle.Y -= BorderWidth / 2;
                rectangle.Width += BorderWidth;
                rectangle.Height += BorderWidth;
                return rectangle;
            }
        }

        internal void ChangeControlPoints(float? x1 = null, float? y1 = null, float? x2 = null, float? y2 = null)
        {
            var rectangle = ControlBounds;
            if (x1.HasValue)
                rectangle.X = x1.Value;
            if (y1.HasValue)
                rectangle.Y = y1.Value;
            if (x2.HasValue)
                rectangle.Width = x2.Value - rectangle.X;
            if (y2.HasValue)
                rectangle.Height = y2.Value - rectangle.Y;

            Path = Math.Path.CreateRectanglePath(rectangle);
        }

        protected internal override Path GetDrawingPath(float dpi = 72)
        {
            if (!IncludeBorder)
                return base.GetDrawingPath(dpi);

            var rect = Rectangle;
            var border = FixedBorderWidth && !Utils.EqualsOfFloatNumbers(dpi, 72) ? BorderWidth * 72f / dpi : BorderWidth;
            rect.Width += border;
            rect.Height += border;

            var path = rect.GetPath();

            if (!Utils.EqualsOfFloatNumbers(dpi, 72))
            {
                using (var matrix = new Matrix())
                {
                    matrix.Scale(dpi / 72, dpi / 72, MatrixOrder.Append);
                    path.ApplyTransform(matrix);
                }
            }

            return path;
        }

        public override void Accept(IVObjectVisitor visitor)
        {
            visitor.Visit(this);
        }
    }
}